Mathlib/CategoryTheory/ConcreteCategory/Basic.lean

```diff
@@ -225,4 +238,187 @@ lemma NatTrans.naturality_apply {C D : Type*} [Category C] [Category D] [HasForg
+  (hom : ∀ {X Y}, (X ⟶ Y) → FC X Y)
+  /-- Convert a bundled function to a morphism of `C`. -/
+  (ofHom : ∀ {X Y}, FC X Y → (X ⟶ Y))
+  (hom_ofHom : ∀ {X Y} (f : FC X Y), hom (ofHom f) = f := by aesop_cat)
+  (ofHom_hom : ∀ {X Y} (f : X ⟶ Y), ofHom (hom f) = f := by aesop_cat)
+  (id_apply : ∀ {X} (x : CC X), hom (𝟙 X) x = x := by aesop_cat)
+  (comp_apply : ∀ {X Y Z} (f : X ⟶ Y) (g : Y ⟶ Z) (x : CC X),
+    hom (f ≫ g) x = hom g (hom f x) := by aesop_cat)
+
+export ConcreteCategory (id_apply comp_apply)
+
+variable {C : Type u} [Category.{v} C] {FC : C → C → Type*} {CC : C → Type w}
+variable [∀ X Y, FunLike (FC X Y) (CC X) (CC Y)]
+variable [ConcreteCategory C FC]
+
+/-- `ToType X` converts the object `X` of the concrete category `C` to a type.
+
+This is an `abbrev` so that instances on `X` (e.g. `Ring`) do not need to be redeclared.
+-/
+@[nolint unusedArguments] -- Need the instance to trigger unification that finds `CC`.
+abbrev ToType [ConcreteCategory C FC] := CC
+
+/-- `ToHom X Y` is the type of (bundled) functions between objects `X Y : C`.
+
+This is an `abbrev` so that instances (e.g. `RingHomClass`) do not need to be redeclared.
+-/
+@[nolint unusedArguments] -- Need the instance to trigger unification that finds `FC`.
+abbrev ToHom [ConcreteCategory C FC] := FC
+
+namespace ConcreteCategory
+
+attribute [simp] id_apply comp_apply
+
+/-- We can apply morphisms of concrete categories by first casting them down
+to the base functions.
+-/
+instance {X Y : C} : CoeFun (X ⟶ Y) (fun _ ↦ ToType X → ToType Y) where
+  coe f := hom f
+
+/--
+`ConcreteCategory.hom` bundled as an `Equiv`.
+-/
+def homEquiv {X Y : C} : (X ⟶ Y) ≃ ToHom X Y where
+  toFun := hom
+  invFun := ofHom
+  left_inv := ofHom_hom
+  right_inv := hom_ofHom
+
+lemma hom_bijective {X Y : C} : Function.Bijective (hom : (X ⟶ Y) → ToHom X Y) :=
+  homEquiv.bijective
+
+lemma hom_injective {X Y : C} : Function.Injective (hom : (X ⟶ Y) → ToHom X Y) :=
+  hom_bijective.injective
+
+/-- In any concrete category, we can test equality of morphisms by pointwise evaluations. -/
+@[ext] lemma ext {X Y : C} {f g : X ⟶ Y} (h : hom f = hom g) : f = g :=
+  hom_injective h
+
+lemma coe_ext {X Y : C} {f g : X ⟶ Y} (h : ⇑(hom f) = ⇑(hom g)) : f = g :=
+  ext (DFunLike.coe_injective h)
+
+lemma ext_apply {X Y : C} {f g : X ⟶ Y} (h : ∀ x, f x = g x) : f = g :=
+  ext (DFunLike.ext _ _ h)
+
+/-- A concrete category comes with a forgetful functor to `Type`.
+
+Warning: because of the way that `ConcreteCategory` and `HasForget` are set up, we can't make
+`forget Type` reducibly defeq to the identity functor. -/
+instance toHasForget : HasForget C where
+  forget.obj := ToType
+  forget.map f := ⇑(hom f)
+  forget_faithful.map_injective h := coe_ext h
+
+end ConcreteCategory
+
+theorem forget_obj (X : C) : (forget C).obj X = ToType X := rfl
+
+@[simp]
+theorem ConcreteCategory.forget_map_eq_coe {X Y : C} (f : X ⟶ Y) : (forget C).map f = f := rfl
```

Comment:
Would you be willing to wrap all the `rfl`'s with `with_reducible_and_instances`? This is more psychological when reading the code than anything.